pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go

```diff
@@ -607,29 +611,37 @@ func (p *processorImpl) makeStateConsistentRaftMuLocked(
 		p.term = state.Term
 	}
 	replicasChanged := p.desc.replicasChanged
-	if replicasChanged {
-		p.desc.replicasChanged = false
-	}
-	if !replicasChanged && state.Leader == p.leaderID && state.Leaseholder == p.leaseholderID &&
-		(p.leader.rc == nil || !termChanged) && !force {
+	leaseChanged := state.Leaseholder != p.leaseholderID
+
+	// Detect when we leave or enter leadership. Note that leftLeader and
+	// becameLeader can be true simultaneously, meaning that we left being leader
+	// of the previous term, and became the leader of state.Term.
+	leftLeader := p.leader.rc != nil && (termChanged || state.State != raft.StateLeader)
```

Comment:
Perhaps it's better to store the `raft.StateType` (or `isLeader` bool) in the `processorImpl` than rely on `p.leader.rc != nil` here, since it has v1/v2 semantics that get on the way.